cpp/harvest/marc_issn_lookup.cc

```diff
@@ -99,86 +152,123 @@ void UpdateSubfield(MARC::Subfields &subfields, const SubFieldInfo &sub_field_in
+            restart = false;
+
+            for (unsigned j = i + 1; j < tmp_cache.size(); j++) {
+                if (IsInISSNs(content.issns_, tmp_cache[j].issns_)) {
+                    UpdateSubFieldAndCombineIssn(&content, tmp_cache[j]);
+                    iter = tmp_cache.begin() + j;
+                    tmp_cache.erase(iter);
+                    restart = true;
+                    break;
+                }
             }
+
+        } while (restart);
+        new_journal_cache.emplace_back(content);
+        restart = false;
     }
-    LOG_INFO("Online (valid): " + std::to_string(online_valid) + ", " + "Online (invalid): " + std::to_string(online_invalid) + ", "
-             + "Print (valid): " + std::to_string(print_valid) + ", " + "Print (invalid): " + std::to_string(print_invalid));
+
+    return new_journal_cache;
 }
 
 
-std::map<std::string, SubFieldInfo> BuildJournalCache(const std::string &input_journal_filename) {
-    std::map<std::string, SubFieldInfo> journal_cache;
+std::vector<SubFieldInfo> BuildJournalCache(const std::string &input_journal_filename) {
+    std::vector<SubFieldInfo> journal_cache;
     auto input_journal_file(MARC::Reader::Factory(input_journal_filename));
-    int record_counter(0);
 
+    std::cout << "building..." << std::endl;
     while (MARC::Record record = input_journal_file->read()) {
         SubFieldInfo sub_field_info_of_record(record);
-
-        // if sub_info is exist in the cache
-        if (journal_cache.find(sub_field_info_of_record.x_) != journal_cache.end()) {
-            if (journal_cache[sub_field_info_of_record.x_].is_online_) {
-                if (sub_field_info_of_record.is_online_)
-                    ++journal_cache[sub_field_info_of_record.x_].online_version_counter_;
-                else
-                    ++journal_cache[sub_field_info_of_record.x_].printed_version_counter_;
-            } else {
-                if (sub_field_info_of_record.is_online_) {
-                    sub_field_info_of_record.printed_version_counter_ = journal_cache[sub_field_info_of_record.x_].printed_version_counter_;
-                    journal_cache[sub_field_info_of_record.x_] = sub_field_info_of_record;
-                    ++journal_cache[sub_field_info_of_record.x_].online_version_counter_;
-                } else
-                    ++journal_cache[sub_field_info_of_record.x_].printed_version_counter_;
+        bool exist_in_journal_cache(false);
+
+        for (auto &elemt : journal_cache) {
+            if (IsInISSNs(sub_field_info_of_record.issns_, elemt.issns_)) {
+                exist_in_journal_cache = true;
+                if (elemt.is_online_) {
+                    if (sub_field_info_of_record.is_online_) {
+                        InsertIfNotExist(sub_field_info_of_record.issns_, &elemt.issns_);
+                        elemt.is_valid_ = false;
+                    } else
+                        InsertIfNotExist(sub_field_info_of_record.issns_, &elemt.issns_);
```

Comment:
This can be simplified even more => if you call InsertIfNotExist in if AND else you might move it 1 level higher as well and remove the empty else branch entirely.